ambry-messageformat/src/main/java/com/github/ambry/messageformat/MessageFormatSend.java

```diff
@@ -118,57 +119,49 @@ private void fetchDataFromReadSet() throws MessageFormatException {
       logger.trace("Calculate offsets of messages for one partition, MessageFormatFlag : {} number of messages : {}",
           flag, messageCount);
       for (int i = 0; i < messageCount; i++) {
-        if (flag == MessageFormatFlags.All) {
+        if (flag == MessageFormatFlags.All || flag == MessageFormatFlags.Blob) {
           // just copy over the total size and use relative offset to be 0
           // We do not have to check any version in this case as we dont
           // have to read any data to deserialize anything.
-          sendInfoList.add(i, new SendInfo(0, readSet.sizeInBytes(i)));
-          messageMetadataList.add(i, null);
-          totalSizeToWrite += readSet.sizeInBytes(i);
           readSet.doPrefetch(i, 0, readSet.sizeInBytes(i));
+
+          if (flag == MessageFormatFlags.All) {
+            sendInfoList.add(i, new SendInfo(0, readSet.sizeInBytes(i)));
+            messageMetadataList.add(i, null);
+            totalSizeToWrite += readSet.sizeInBytes(i);
+          } else if (flag == MessageFormatFlags.Blob) {
+            ByteBuf blobAll = readSet.getPrefetchedData(i);
```

Comment:
I am a bit worried about the using prefetchedData here, without adding it to the `dataFromReadSet` list.

If there is any exception while parsing the header, this `ByteBuf` will be leaked. We should probably add the prefetchedData ByteBuf to the `dataFromReadSet` list before using it to parse the header.

Reply:
`dataFromReadSet.add(readSet.getPrefetchedData(i));` also covers this case. (It's after the if-else for each message). no?

Reply:
it does, but what if at line 136, parseHeaderAndVerifyStoreKey failed with an exception, then this ByteBuf will not be released.

Reply:
Got it. changes made.